
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>TCR-DeepInsight Vignette &#8212; TCR paper notebook</title>
    
  <link href="_static/css/theme.css" rel="stylesheet">
  <link href="_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/clipboard.min.js"></script>
    <script src="_static/copybutton.js"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="_static/sphinx-book-theme.d59cb220de22ca1c485ebbdc042f0030.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="Import Basic Packages and Definitions" href="huARdb_TCR_paper_notebook.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">TCR paper notebook</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="intro.html">
   Welcome to the huARdb TCR analysis notebook
  </a>
 </li>
</ul>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="huARdb_TCR_paper_notebook.html">
   Import Basic Packages and Definitions
  </a>
 </li>
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   TCR-DeepInsight Vignette
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="_sources/huARdb_TCR_DeepInsight_notebook.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
                onclick="printPdf(this)" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/WanluLiuLab/TCR-DeepInsight"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/WanluLiuLab/TCR-DeepInsight/issues/new?title=Issue%20on%20page%20%2FhuARdb_TCR_DeepInsight_notebook.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/WanluLiuLab/TCR-DeepInsight/master?urlpath=tree/docs/huARdb_TCR_DeepInsight_notebook.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show noprint">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   TCR-DeepInsight Vignette
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#training-new-reference-datasets">
   Training new reference datasets
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#training-gex-reference">
     Training GEX reference
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#training-tcr-reference">
     Training TCR reference
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#cluster-tcrs-in-reference-datasets">
     Cluster TCRs in reference datasets
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cluster-tcrs-in-querying-datasets-from-reference-datasets">
   Cluster TCRs in querying datasets from reference datasets
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
            <!-- Table of contents that is only displayed when printing the page -->
            <div id="jb-print-docs-body" class="onlyprint">
                <h1>TCR-DeepInsight Vignette</h1>
                <!-- Table of contents -->
                <div id="print-main-content">
                    <div id="jb-print-toc">
                        
                        <div>
                            <h2> Contents </h2>
                        </div>
                        <nav aria-label="Page">
                            <ul class="visible nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   TCR-DeepInsight Vignette
  </a>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#training-new-reference-datasets">
   Training new reference datasets
  </a>
  <ul class="visible nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#training-gex-reference">
     Training GEX reference
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#training-tcr-reference">
     Training TCR reference
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#cluster-tcrs-in-reference-datasets">
     Cluster TCRs in reference datasets
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#cluster-tcrs-in-querying-datasets-from-reference-datasets">
   Cluster TCRs in querying datasets from reference datasets
  </a>
 </li>
</ul>

                        </nav>
                    </div>
                </div>
            </div>
            
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="tcr-deepinsight-vignette">
<h1>TCR-DeepInsight Vignette<a class="headerlink" href="#tcr-deepinsight-vignette" title="Permalink to this headline">¶</a></h1>
<p>The emergence of single-cell immune profiling technology has led to the production of a large amount of data on single-cell gene expression (GEX) and T cell receptor (TCR), which has great potential for studying TCR biology and identifying effective TCRs. However, one of the major challenges is the lack of a reference atlas that provides easy access to these datasets. On the other hand, the use of TCR engineering in disease immunotherapy is rapidly advancing, and single-cell immune profiling data can be a valuable resource for identifying functional TCRs. Nevertheless, the lack of efficient computational tools to integrate and identify functional TCRs is a significant obstacle in this field.</p>
<p>The TCR-DeepInsight is a module that can perform <strong>GEX</strong> / <strong>TCR</strong> joint analysis.</p>
<div class="alert alert-info">
<p><strong>Note</strong></p>
<p>In Jupyter notebooks and lab, you can see the documentation for a python function by hitting <code class="docutils literal notranslate"><span class="pre">SHIFT</span> <span class="pre">+</span> <span class="pre">TAB</span></code>. Hit it twice to expand the view.</p>
</div><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">tcr_deep_insight</span> <span class="k">as</span> <span class="nn">tdi</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="training-new-reference-datasets">
<h1>Training new reference datasets<a class="headerlink" href="#training-new-reference-datasets" title="Permalink to this headline">¶</a></h1>
<div class="section" id="training-gex-reference">
<h2>Training GEX reference<a class="headerlink" href="#training-gex-reference" title="Permalink to this headline">¶</a></h2>
<p>The GEX data is stored in an .h5ad file, which contains <strong>raw gene expression matrix</strong> and meta informations including <strong>study_name</strong>, <strong>sample_name</strong>, and other annotations</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gex_reference_adata</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">human_gex_reference_v1</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gex_reference_adata</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AnnData object with n_obs × n_vars = 1017877 × 6000
    obs: &#39;IR_VJ_1_junction_aa&#39;, &#39;IR_VDJ_1_junction_aa&#39;, &#39;IR_VJ_1_v_call&#39;, &#39;IR_VJ_1_j_call&#39;, &#39;IR_VDJ_1_v_call&#39;, &#39;IR_VDJ_1_j_call&#39;, &#39;individual&#39;, &#39;cell_subtype&#39;, &#39;disease&#39;, &#39;disease_type&#39;, &#39;tcr&#39;, &#39;cell_type_1&#39;, &#39;cell_type_2&#39;, &#39;cell_type_3&#39;, &#39;study_name&#39;, &#39;sample_name&#39;, &#39;disease_type_1&#39;, &#39;disease_type_2&#39;
    obsm: &#39;X_gex&#39;
</pre></div>
</div>
</div>
</div>
<p>Here we use the reference GEX data from the paper, which contains 1,017,877 T cells from various studies and diseases</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gex_reference_adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>IR_VJ_1_junction_aa</th>
      <th>IR_VDJ_1_junction_aa</th>
      <th>IR_VJ_1_v_call</th>
      <th>IR_VJ_1_j_call</th>
      <th>IR_VDJ_1_v_call</th>
      <th>IR_VDJ_1_j_call</th>
      <th>individual</th>
      <th>cell_subtype</th>
      <th>disease</th>
      <th>disease_type</th>
      <th>tcr</th>
      <th>cell_type_1</th>
      <th>cell_type_2</th>
      <th>cell_type_3</th>
      <th>study_name</th>
      <th>sample_name</th>
      <th>disease_type_1</th>
      <th>disease_type_2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACCTGAGAAACCAT-Wen_2020-HC1</th>
      <td>CAVRDNTNAGKSTF</td>
      <td>CASSKHPDRYEQYF</td>
      <td>TRAV1-2</td>
      <td>TRAJ27</td>
      <td>TRBV7-9</td>
      <td>TRBJ2-7</td>
      <td>HC1</td>
      <td>NaN</td>
      <td>Healthy</td>
      <td>Healthy</td>
      <td>CAVRDNTNAGKSTF=CASSKHPDRYEQYF=TRAV1-2=TRAJ27=T...</td>
      <td>Naive CD4</td>
      <td>Naive CD4</td>
      <td>CD4</td>
      <td>Wen_2020</td>
      <td>HC1</td>
      <td>Healthy</td>
      <td>Healthy</td>
    </tr>
    <tr>
      <th>AAACCTGAGAAACCAT-Yost_2019-su012_post_Tcell</th>
      <td>CAALFSDGQKLLF</td>
      <td>CASSLRQGAGSNQPQHF</td>
      <td>TRAV39</td>
      <td>TRAJ16</td>
      <td>TRBV5-5</td>
      <td>TRBJ1-5</td>
      <td>su012</td>
      <td>NaN</td>
      <td>Basal cell carinoma tumor</td>
      <td>Solid tumor</td>
      <td>CAALFSDGQKLLF=CASSLRQGAGSNQPQHF=TRAV39=TRAJ16=...</td>
      <td>CD4</td>
      <td>CD4</td>
      <td>CD4</td>
      <td>Yost_2019</td>
      <td>su012_post_Tcell</td>
      <td>Solid tumor</td>
      <td>Solid tumor</td>
    </tr>
    <tr>
      <th>AAACCTGAGAAAGTGG-Gao_2020-HD5</th>
      <td>CAASRGDNDMRF</td>
      <td>CASSLRATDTQYF</td>
      <td>TRAV13-1</td>
      <td>TRAJ43</td>
      <td>TRBV7-9</td>
      <td>TRBJ2-3</td>
      <td>HD5</td>
      <td>NaN</td>
      <td>Healthy</td>
      <td>Healthy</td>
      <td>CAASRGDNDMRF=CASSLRATDTQYF=TRAV13-1=TRAJ43=TRB...</td>
      <td>CD4</td>
      <td>CD4</td>
      <td>CD4</td>
      <td>Gao_2020</td>
      <td>HD5</td>
      <td>Healthy</td>
      <td>Healthy</td>
    </tr>
    <tr>
      <th>AAACCTGAGAAAGTGG-Gao_2020-UPN18_post</th>
      <td>CAARVQGAQKLVF</td>
      <td>CAWSVPRGLDEQYF</td>
      <td>TRAV29DV5</td>
      <td>TRAJ54</td>
      <td>TRBV30</td>
      <td>TRBJ2-7</td>
      <td>UPN18</td>
      <td>NaN</td>
      <td>Large granular lymphocyte leukemia</td>
      <td>T-LGLL</td>
      <td>CAARVQGAQKLVF=CAWSVPRGLDEQYF=TRAV29DV5=TRAJ54=...</td>
      <td>CD4</td>
      <td>CD4</td>
      <td>CD4</td>
      <td>Gao_2020</td>
      <td>UPN18_post</td>
      <td>T-LGLL</td>
      <td>T-LGLL</td>
    </tr>
    <tr>
      <th>AAACCTGAGAAAGTGG-Krishna_2021-IpiNivo_Mixed_PBMC</th>
      <td>CAASEGAGGFKTIF</td>
      <td>CASRSDRGPNQPQHF</td>
      <td>TRAV29DV5</td>
      <td>TRAJ9</td>
      <td>TRBV28</td>
      <td>TRBJ1-5</td>
      <td>IpiNivo_Mixed</td>
      <td>NaN</td>
      <td>Clear cell renal cell carcinoma</td>
      <td>Solid tumor</td>
      <td>CAASEGAGGFKTIF=CASRSDRGPNQPQHF=TRAV29DV5=TRAJ9...</td>
      <td>CD4</td>
      <td>CD4</td>
      <td>CD4</td>
      <td>Krishna_2021</td>
      <td>IpiNivo_Mixed_PBMC</td>
      <td>Solid tumor</td>
      <td>Solid tumor</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TTTGGTTTCATCTGCC-Simone_2021-AS02_SF</th>
      <td>CLPVSGWEGGGNKLTF</td>
      <td>CASSLDSPEQYF</td>
      <td>TRAV26-1</td>
      <td>TRAJ10</td>
      <td>TRBV5-4</td>
      <td>TRBJ2-7</td>
      <td>AS02</td>
      <td>GZMK+ Tex</td>
      <td>Ankylosing spondylitis</td>
      <td>Inflammation</td>
      <td>CLPVSGWEGGGNKLTF=CASSLDSPEQYF=TRAV26-1=TRAJ10=...</td>
      <td>CD8</td>
      <td>CD8</td>
      <td>CD8</td>
      <td>Simone_2021</td>
      <td>AS02_SF</td>
      <td>Inflammation</td>
      <td>Ankylosing spondylitis</td>
    </tr>
    <tr>
      <th>TTTGTCAAGCGTAATA-Simone_2021-AS02_SF</th>
      <td>CAMKGNAGNNRKLIW</td>
      <td>CASSWTGVISGNTIYF</td>
      <td>TRAV14DV4</td>
      <td>TRAJ38</td>
      <td>TRBV5-5</td>
      <td>TRBJ1-3</td>
      <td>AS02</td>
      <td>Cycling T</td>
      <td>Ankylosing spondylitis</td>
      <td>Inflammation</td>
      <td>CAMKGNAGNNRKLIW=CASSWTGVISGNTIYF=TRAV14DV4=TRA...</td>
      <td>Cycling</td>
      <td>Cycling</td>
      <td>Undefined</td>
      <td>Simone_2021</td>
      <td>AS02_SF</td>
      <td>Inflammation</td>
      <td>Ankylosing spondylitis</td>
    </tr>
    <tr>
      <th>TTTGTCAAGCTGTCTA-Simone_2021-AS02_SF</th>
      <td>CAERDGYSTLTF</td>
      <td>CSARARSYNEQFF</td>
      <td>TRAV5</td>
      <td>TRAJ11</td>
      <td>TRBV20-1</td>
      <td>TRBJ2-1</td>
      <td>AS02</td>
      <td>GZMK+ Tex</td>
      <td>Ankylosing spondylitis</td>
      <td>Inflammation</td>
      <td>CAERDGYSTLTF=CSARARSYNEQFF=TRAV5=TRAJ11=TRBV20...</td>
      <td>CD8</td>
      <td>CD8</td>
      <td>CD8</td>
      <td>Simone_2021</td>
      <td>AS02_SF</td>
      <td>Inflammation</td>
      <td>Ankylosing spondylitis</td>
    </tr>
    <tr>
      <th>TTTGTCACATCACGTA-Simone_2021-AS02_SF</th>
      <td>CVVRKPQDYNNARLMF</td>
      <td>CASSLGGSYNEQFF</td>
      <td>TRAV8-2</td>
      <td>TRAJ31</td>
      <td>TRBV27</td>
      <td>TRBJ2-1</td>
      <td>AS02</td>
      <td>Cycling T</td>
      <td>Ankylosing spondylitis</td>
      <td>Inflammation</td>
      <td>CVVRKPQDYNNARLMF=CASSLGGSYNEQFF=TRAV8-2=TRAJ31...</td>
      <td>CD8</td>
      <td>CD8</td>
      <td>CD8</td>
      <td>Simone_2021</td>
      <td>AS02_SF</td>
      <td>Inflammation</td>
      <td>Ankylosing spondylitis</td>
    </tr>
    <tr>
      <th>TTTGTCAGTTGCGTTA-Simone_2021-AS02_SF</th>
      <td>CAMRESNTGRRALTF</td>
      <td>CASSSDRDRGREQYF</td>
      <td>TRAV14DV4</td>
      <td>TRAJ5</td>
      <td>TRBV7-9</td>
      <td>TRBJ2-7</td>
      <td>AS02</td>
      <td>GZMK+ Tex</td>
      <td>Ankylosing spondylitis</td>
      <td>Inflammation</td>
      <td>CAMRESNTGRRALTF=CASSSDRDRGREQYF=TRAV14DV4=TRAJ...</td>
      <td>CD8</td>
      <td>CD8</td>
      <td>CD8</td>
      <td>Simone_2021</td>
      <td>AS02_SF</td>
      <td>Inflammation</td>
      <td>Ankylosing spondylitis</td>
    </tr>
  </tbody>
</table>
<p>1017877 rows × 18 columns</p>
</div></div></div>
</div>
<p>The VAEModel for GEX can be easily set up by the following code. The <strong>sample_name</strong> is set for the batch key for correcting batch effects.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vae_model</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">VAEModel</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">gex_reference_adata</span><span class="p">,</span> 
    <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;sample_name&quot;</span><span class="p">,</span> 
    <span class="n">device</span><span class="o">=</span><span class="s1">&#39;cuda:1&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Thu May  4 14:15:46 2023	Initializing dataset into memory
Thu May  4 14:15:47 2023	warning: the provided n_batch=0 does not match the number of batch in the adata.
Thu May  4 14:15:47 2023	setting n_batch to 437
</pre></div>
</div>
</div>
</div>
<p>Fitting the VAE model is ultra-fast with GPU support. We use only 16 epoch to train the VAE model as the number of cells is very large.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">vae_model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span>
    <span class="n">max_epoch</span><span class="o">=</span><span class="mi">16</span><span class="p">,</span>
    <span class="n">lr</span><span class="o">=</span><span class="mf">1e-4</span><span class="p">,</span>
    <span class="n">n_per_batch</span><span class="o">=</span><span class="mi">256</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 16: 100%|██████████████████████████████████████| 16/16 [14:44&lt;00:00, 55.28s/it, loss=3.36e+05, reconst=3.33e+05, kldiv=2.96e+03]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">The</span> <span class="n">latent</span> <span class="n">embedding</span> <span class="kn">from</span> <span class="nn">the</span> <span class="n">VAE</span> <span class="n">model</span> <span class="n">can</span> <span class="n">be</span> <span class="n">easily</span> <span class="n">obtained</span> <span class="ow">and</span> <span class="n">projected</span> <span class="n">to</span> <span class="mi">2</span><span class="n">D</span> <span class="n">dimension</span><span class="o">.</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gex_reference_adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_embedding&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">X_embedding</span> <span class="o">=</span> <span class="n">vae_model</span><span class="o">.</span><span class="n">get_latent_embedding</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">umap</span>
<span class="n">gex_reference_adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_umap&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="o">.</span><span class="n">UMAP</span><span class="p">(</span><span class="n">min_dist</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span><span class="o">.</span><span class="n">fit_transform</span><span class="p">(</span><span class="n">X_embedding</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>
<span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">gex_reference_adata</span><span class="p">,</span> 
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cell_type_1&#39;</span><span class="p">,</span> 
    <span class="n">palette</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palette</span><span class="o">.</span><span class="n">reannotated_prediction_palette</span><span class="p">,</span>
    <span class="n">ax</span><span class="o">=</span><span class="n">ax</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/huARdb_TCR_DeepInsight_notebook_19_0.png" src="_images/huARdb_TCR_DeepInsight_notebook_19_0.png" />
</div>
</div>
</div>
<div class="section" id="training-tcr-reference">
<h2>Training TCR reference<a class="headerlink" href="#training-tcr-reference" title="Permalink to this headline">¶</a></h2>
<p>Similar to the GEX data, we store the TCR data in a .h5ad file, which contains the full-length
TCR sequence. We aggregate the TCR sequence by <strong>individual</strong> to obtain unique clonotypes.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_reference_adata</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;./tcr_deep_insight/data/reference/human_tcr_reference_v1.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>TRAV</th>
      <th>TRAJ</th>
      <th>TRBV</th>
      <th>TRBJ</th>
      <th>CDR3a</th>
      <th>CDR3b</th>
      <th>individual</th>
      <th>disease</th>
      <th>disease_type</th>
      <th>cell_subtype</th>
      <th>tcr</th>
      <th>disease_type_1</th>
      <th>disease_type_2</th>
      <th>CDR3a_mr</th>
      <th>CDR3b_mr</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CAAAAAGKSTF-CASSQGDSSYEQYF-TRAV12-2-TRAJ27-TRBV7-4-TRBJ2-7-IpiNivo_Resistant</th>
      <td>TRAV12-2</td>
      <td>TRAJ27</td>
      <td>TRBV7-4</td>
      <td>TRBJ2-7</td>
      <td>CAAAAAGKSTF</td>
      <td>CASSQGDSSYEQYF</td>
      <td>IpiNivo_Resistant</td>
      <td>Clear cell renal cell carcinoma</td>
      <td>Solid tumor</td>
      <td>CD8</td>
      <td>CAAAAAGKSTF=CASSQGDSSYEQYF=TRAV12-2=TRAJ27=TRB...</td>
      <td>Solid tumor</td>
      <td>Solid tumor</td>
      <td>AAA</td>
      <td>QGDS</td>
    </tr>
    <tr>
      <th>CAAAAAGNKLTF-CASSGDPIVGEQFF-TRAV13-1-TRAJ17-TRBV5-6-TRBJ2-1-PBMC_C_HD2</th>
      <td>TRAV13-1</td>
      <td>TRAJ17</td>
      <td>TRBV5-6</td>
      <td>TRBJ2-1</td>
      <td>CAAAAAGNKLTF</td>
      <td>CASSGDPIVGEQFF</td>
      <td>PBMC_C_HD2</td>
      <td>Healthy</td>
      <td>COVID-19</td>
      <td>CD4</td>
      <td>CAAAAAGNKLTF=CASSGDPIVGEQFF=TRAV13-1=TRAJ17=TR...</td>
      <td>Healthy</td>
      <td>Healthy</td>
      <td>A</td>
      <td>GDPIVG</td>
    </tr>
    <tr>
      <th>CAAAAAGNKLTF-CSAREQGWGEQFF-TRAV13-1-TRAJ17-TRBV20-1-TRBJ2-1-PBMC_P5</th>
      <td>TRAV13-1</td>
      <td>TRAJ17</td>
      <td>TRBV20-1</td>
      <td>TRBJ2-1</td>
      <td>CAAAAAGNKLTF</td>
      <td>CSAREQGWGEQFF</td>
      <td>PBMC_P5</td>
      <td>MIS-C</td>
      <td>COVID-19</td>
      <td>CD4</td>
      <td>CAAAAAGNKLTF=CSAREQGWGEQFF=TRAV13-1=TRAJ17=TRB...</td>
      <td>COVID-19</td>
      <td>COVID-19</td>
      <td>A</td>
      <td>EQGWG</td>
    </tr>
    <tr>
      <th>CAAAAAGRRALTF-CSARGGWGGITGELFF-TRAV12-2-TRAJ5-TRBV20-1-TRBJ2-2-IpiNivo_Complete</th>
      <td>TRAV12-2</td>
      <td>TRAJ5</td>
      <td>TRBV20-1</td>
      <td>TRBJ2-2</td>
      <td>CAAAAAGRRALTF</td>
      <td>CSARGGWGGITGELFF</td>
      <td>IpiNivo_Complete</td>
      <td>Clear cell renal cell carcinoma</td>
      <td>Solid tumor</td>
      <td>CD4</td>
      <td>CAAAAAGRRALTF=CSARGGWGGITGELFF=TRAV12-2=TRAJ5=...</td>
      <td>Solid tumor</td>
      <td>Solid tumor</td>
      <td>AAAA</td>
      <td>GGWGGI</td>
    </tr>
    <tr>
      <th>CAAAAARGSYIPTF-CASSLSDYGYTF-TRAV21-TRAJ6-TRBV5-6-TRBJ1-2-C-39</th>
      <td>TRAV21</td>
      <td>TRAJ6</td>
      <td>TRBV5-6</td>
      <td>TRBJ1-2</td>
      <td>CAAAAARGSYIPTF</td>
      <td>CASSLSDYGYTF</td>
      <td>C-39</td>
      <td>COVID-19</td>
      <td>COVID-19</td>
      <td>CD8</td>
      <td>CAAAAARGSYIPTF=CASSLSDYGYTF=TRAV21=TRAJ6=TRBV5...</td>
      <td>COVID-19</td>
      <td>COVID-19</td>
      <td>AAAAR</td>
      <td>SD</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>CYVGGTSYGKLTF-CASSLEREIYGYTF-TRAV8-3-TRAJ52-TRBV5-4-TRBJ1-2-HD4</th>
      <td>TRAV8-3</td>
      <td>TRAJ52</td>
      <td>TRBV5-4</td>
      <td>TRBJ1-2</td>
      <td>CYVGGTSYGKLTF</td>
      <td>CASSLEREIYGYTF</td>
      <td>HD4</td>
      <td>Healthy</td>
      <td>Healthy</td>
      <td>CD4</td>
      <td>CYVGGTSYGKLTF=CASSLEREIYGYTF=TRAV8-3=TRAJ52=TR...</td>
      <td>Healthy</td>
      <td>Healthy</td>
      <td></td>
      <td>EREI</td>
    </tr>
    <tr>
      <th>CYYDKVIF-CSARAIAGGPPETQYF-TRAV12-3-TRAJ50-TRBV20-1-TRBJ2-5-NPC_SC_1813</th>
      <td>TRAV12-3</td>
      <td>TRAJ50</td>
      <td>TRBV20-1</td>
      <td>TRBJ2-5</td>
      <td>CYYDKVIF</td>
      <td>CSARAIAGGPPETQYF</td>
      <td>NPC_SC_1813</td>
      <td>Nasopharyngeal carcinoma</td>
      <td>Solid tumor</td>
      <td>CD4</td>
      <td>CYYDKVIF=CSARAIAGGPPETQYF=TRAV12-3=TRAJ50=TRBV...</td>
      <td>Solid tumor</td>
      <td>Solid tumor</td>
      <td>Y</td>
      <td>AIAGGPP</td>
    </tr>
    <tr>
      <th>CYYPYKLSF-CASSIGGTEAFF-TRAV12-1-TRAJ20-TRBV19-TRBJ1-1-IpiNivo_Resistant</th>
      <td>TRAV12-1</td>
      <td>TRAJ20</td>
      <td>TRBV19</td>
      <td>TRBJ1-1</td>
      <td>CYYPYKLSF</td>
      <td>CASSIGGTEAFF</td>
      <td>IpiNivo_Resistant</td>
      <td>Clear cell renal cell carcinoma</td>
      <td>Solid tumor</td>
      <td>CD8</td>
      <td>CYYPYKLSF=CASSIGGTEAFF=TRAV12-1=TRAJ20=TRBV19=...</td>
      <td>Solid tumor</td>
      <td>Solid tumor</td>
      <td>YYP</td>
      <td>GG</td>
    </tr>
    <tr>
      <th>CYYSATGNQFYF-CASSPGGLVNEQFF-TRAV24-TRAJ49-TRBV27-TRBJ2-1-UPN10</th>
      <td>TRAV24</td>
      <td>TRAJ49</td>
      <td>TRBV27</td>
      <td>TRBJ2-1</td>
      <td>CYYSATGNQFYF</td>
      <td>CASSPGGLVNEQFF</td>
      <td>UPN10</td>
      <td>Large granular lymphocyte leukemia</td>
      <td>T-LGLL</td>
      <td>CD4</td>
      <td>CYYSATGNQFYF=CASSPGGLVNEQFF=TRAV24=TRAJ49=TRBV...</td>
      <td>T-LGLL</td>
      <td>T-LGLL</td>
      <td>YYSA</td>
      <td>PGGLV</td>
    </tr>
    <tr>
      <th>CYYSGGYQKVTF-CAISESKSGAEKLFF-TRAV25-TRAJ13-TRBV10-3-TRBJ1-4-su001</th>
      <td>TRAV25</td>
      <td>TRAJ13</td>
      <td>TRBV10-3</td>
      <td>TRBJ1-4</td>
      <td>CYYSGGYQKVTF</td>
      <td>CAISESKSGAEKLFF</td>
      <td>su001</td>
      <td>Basal cell carinoma tumor</td>
      <td>Solid tumor</td>
      <td>Treg</td>
      <td>CYYSGGYQKVTF=CAISESKSGAEKLFF=TRAV25=TRAJ13=TRB...</td>
      <td>Solid tumor</td>
      <td>Solid tumor</td>
      <td>YY</td>
      <td>SKSGA</td>
    </tr>
  </tbody>
</table>
<p>635393 rows × 15 columns</p>
</div></div></div>
</div>
<p>We implemented a tokenizer for the full-length TCR sequence.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_tokenizer</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">TCRabTokenizer</span><span class="p">(</span>
    <span class="n">tra_max_length</span><span class="o">=</span><span class="mi">48</span><span class="p">,</span> 
    <span class="n">trb_max_length</span><span class="o">=</span><span class="mi">48</span><span class="p">,</span>
    <span class="n">species</span><span class="o">=</span><span class="s1">&#39;human&#39;</span> <span class="c1"># or &#39;mouse&#39;</span>
<span class="p">)</span>
<span class="n">tcr_dataset</span> <span class="o">=</span> <span class="n">tcr_tokenizer</span><span class="o">.</span><span class="n">to_dataset</span><span class="p">(</span>
   <span class="n">ids</span><span class="o">=</span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">,</span>
   <span class="n">alpha_v_genes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;TRAV&#39;</span><span class="p">]),</span>
   <span class="n">alpha_j_genes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;TRAJ&#39;</span><span class="p">]),</span>
   <span class="n">beta_v_genes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;TRBV&#39;</span><span class="p">]),</span>
   <span class="n">beta_j_genes</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;TRBJ&#39;</span><span class="p">]),</span>
   <span class="n">alpha_chains</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;CDR3a&#39;</span><span class="p">]),</span>
   <span class="n">beta_chains</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;CDR3b&#39;</span><span class="p">]),</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>If the dataset is for training, we can split the dataset into train and test set.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Train Test Split</span>
<span class="n">tcr_dataset</span> <span class="o">=</span> <span class="n">tcr_dataset</span><span class="p">[</span><span class="s1">&#39;train&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">train_test_split</span><span class="p">(</span><span class="mf">0.05</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Parameter &#39;generator&#39;=Generator(PCG64) of the transform datasets.arrow_dataset.Dataset.train_test_split couldn&#39;t be hashed properly, a random hash was used instead. Make sure your transforms and parameters are serializable with pickle or dill for the dataset fingerprinting and caching to work. If you reuse this transform, the caching mechanism will consider it to be different from the previous calls and recompute everything. This warning is only showed once. Subsequent hashing failures won&#39;t be showed.
</pre></div>
</div>
</div>
</div>
<p>The dataset can be saved to disk</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_dataset</span><span class="o">.</span><span class="n">save_to_disk</span><span class="p">(</span><span class="s2">&quot;./tcr_deep_insight/data/datasets/human_tcr_v1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">datasets</span>
<span class="n">tcr_dataset</span> <span class="o">=</span> <span class="n">datasets</span><span class="o">.</span><span class="n">load_from_disk</span><span class="p">(</span><span class="s2">&quot;./tcr_deep_insight/data/datasets/human_tcr_v1&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_model</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">TCRabModel</span><span class="p">(</span>
    <span class="n">tdi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">get_human_config</span><span class="p">(),</span>
    <span class="n">labels_number</span><span class="o">=</span><span class="mi">1</span>
<span class="p">)</span><span class="o">.</span><span class="n">to</span><span class="p">(</span><span class="s2">&quot;cuda:1&quot;</span><span class="p">)</span>

<span class="n">tcr_collator</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">TCRabCollator</span><span class="p">(</span><span class="mi">48</span><span class="p">,</span> <span class="mi">48</span><span class="p">,</span> <span class="n">mlm_probability</span><span class="o">=</span><span class="mf">0.15</span><span class="p">)</span>
<span class="n">tcr_model_optimizers</span> <span class="o">=</span> <span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">optim</span><span class="o">.</span><span class="n">AdamW</span><span class="p">(</span><span class="n">tcr_model</span><span class="o">.</span><span class="n">parameters</span><span class="p">(),</span> <span class="n">lr</span><span class="o">=</span><span class="mf">1e-4</span><span class="p">),</span> <span class="kc">None</span><span class="p">)</span>
<span class="n">tcr_model_trainer</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">TCRabTrainer</span><span class="p">(</span>
    <span class="n">tcr_model</span><span class="p">,</span> 
    <span class="n">collator</span><span class="o">=</span><span class="n">tcr_collator</span><span class="p">,</span> 
    <span class="n">train_dataset</span><span class="o">=</span><span class="n">tcr_dataset</span><span class="p">[</span><span class="s1">&#39;train&#39;</span><span class="p">],</span> 
    <span class="n">test_dataset</span><span class="o">=</span><span class="n">tcr_dataset</span><span class="p">[</span><span class="s1">&#39;test&#39;</span><span class="p">],</span> 
    <span class="n">optimizers</span><span class="o">=</span><span class="n">tcr_model_optimizers</span><span class="p">,</span> 
    <span class="n">device</span><span class="o">=</span><span class="s1">&#39;cuda:1&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_model_trainer</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">max_epoch</span><span class="o">=</span><span class="mi">3</span><span class="p">,</span> <span class="n">show_progress_bar</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>  5%|██████████▊                                                                                                                                                                                                      | 3129/60362 [03:53&lt;1:13:05, 13.05it/s, Loss=0.0457, Prediction=0]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">torch</span><span class="o">.</span><span class="n">save</span><span class="p">(</span><span class="n">tcr_model</span><span class="o">.</span><span class="n">state_dict</span><span class="p">(),</span> <span class="s1">&#39;/path/to/saved/checkpoints&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_model</span><span class="o">.</span><span class="n">load_state_dict</span><span class="p">(</span><span class="n">torch</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s2">&quot;./tcr_deep_insight/data/pretrained_weights/human_bert_tcr_768_v1.ckpt&quot;</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>&lt;All keys matched successfully&gt;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_model_trainer</span><span class="o">.</span><span class="n">attach_train_dataset</span><span class="p">(</span><span class="n">tcr_dataset</span><span class="p">[</span><span class="s1">&#39;train&#39;</span><span class="p">])</span>
<span class="n">tcr_model_trainer</span><span class="o">.</span><span class="n">attach_test_dataset</span><span class="p">(</span><span class="n">tcr_dataset</span><span class="p">[</span><span class="s1">&#39;test&#39;</span><span class="p">])</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">all_train_result</span><span class="p">,</span> <span class="n">all_test_result</span> <span class="o">=</span> <span class="n">tcr_model_trainer</span><span class="o">.</span><span class="n">evaluate</span><span class="p">(</span><span class="n">n_per_batch</span><span class="o">=</span><span class="mi">64</span><span class="p">,</span> <span class="n">show_progress</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">Counter</span>

<span class="k">def</span> <span class="nf">FLATTEN</span><span class="p">(</span><span class="n">x</span><span class="p">):</span> 
    <span class="k">return</span> <span class="p">[</span><span class="n">i</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">x</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">s</span><span class="p">]</span>

<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;==== Train ====&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;masked amino acids prediction </span><span class="si">{:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;aa&#39;</span><span class="p">]))[</span><span class="kc">True</span><span class="p">]</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;aa&#39;</span><span class="p">]))))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;masked TRAV prediction </span><span class="si">{:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;av&#39;</span><span class="p">]))[</span><span class="kc">True</span><span class="p">]</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;av&#39;</span><span class="p">]))))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;masked TRBV prediction </span><span class="si">{:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;bv&#39;</span><span class="p">]))[</span><span class="kc">True</span><span class="p">]</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;bv&#39;</span><span class="p">]))))</span>
<span class="nb">print</span><span class="p">(</span><span class="s2">&quot;==== Test ====&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;masked amino acids prediction </span><span class="si">{:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_test_result</span><span class="p">[</span><span class="s1">&#39;aa&#39;</span><span class="p">]))[</span><span class="kc">True</span><span class="p">]</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_test_result</span><span class="p">[</span><span class="s1">&#39;aa&#39;</span><span class="p">]))))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;masked TRAV prediction </span><span class="si">{:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_test_result</span><span class="p">[</span><span class="s1">&#39;av&#39;</span><span class="p">]))[</span><span class="kc">True</span><span class="p">]</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_test_result</span><span class="p">[</span><span class="s1">&#39;av&#39;</span><span class="p">]))))</span>
<span class="nb">print</span><span class="p">(</span><span class="s1">&#39;masked TRBV prediction </span><span class="si">{:.3f}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">Counter</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_test_result</span><span class="p">[</span><span class="s1">&#39;bv&#39;</span><span class="p">]))[</span><span class="kc">True</span><span class="p">]</span> <span class="o">/</span> <span class="nb">len</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_test_result</span><span class="p">[</span><span class="s1">&#39;bv&#39;</span><span class="p">]))))</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>==== Train ====
masked amino acids prediction 0.781
masked TRAV prediction 0.778
masked TRBV prediction 0.671
==== Test ====
masked amino acids prediction 0.784
masked TRAV prediction 0.780
masked TRBV prediction 0.669
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">sklearn</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">labels</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;aa_gt&#39;</span><span class="p">])</span> <span class="o">+</span> <span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;aa_pred&#39;</span><span class="p">]))</span>
<span class="n">cm</span> <span class="o">=</span> <span class="n">sklearn</span><span class="o">.</span><span class="n">metrics</span><span class="o">.</span><span class="n">confusion_matrix</span><span class="p">(</span>
    <span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;aa_gt&#39;</span><span class="p">]),</span> 
    <span class="n">FLATTEN</span><span class="p">(</span><span class="n">all_train_result</span><span class="p">[</span><span class="s1">&#39;aa_pred&#39;</span><span class="p">]),</span> 
    <span class="n">labels</span><span class="o">=</span><span class="n">labels</span>
<span class="p">)</span>
<span class="n">cm</span> <span class="o">=</span> <span class="n">cm</span><span class="p">[:</span><span class="mi">20</span><span class="p">,:][:,:</span><span class="mi">20</span><span class="p">]</span>
<span class="n">cm</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
    <span class="n">cm</span><span class="p">,</span> 
    <span class="n">index</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">_AMINO_ACIDS_INDEX_REVERSE</span><span class="p">[</span><span class="n">x</span><span class="p">],</span> <span class="n">labels</span><span class="p">[:</span><span class="mi">20</span><span class="p">])),</span>
    <span class="n">columns</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">_AMINO_ACIDS_INDEX_REVERSE</span><span class="p">[</span><span class="n">x</span><span class="p">],</span> <span class="n">labels</span><span class="p">[:</span><span class="mi">20</span><span class="p">]))</span>
<span class="p">)</span>
<span class="kn">from</span> <span class="nn">tcr_deep_insight.utils._amino_acids</span> <span class="kn">import</span> <span class="n">_AMINO_ACIDS_INDEX_REVERSE</span>
<span class="n">sns</span><span class="o">.</span><span class="n">clustermap</span><span class="p">(</span><span class="n">cm</span><span class="p">,</span> <span class="n">row_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">col_cluster</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">standard_scale</span><span class="o">=</span><span class="s1">&#39;var&#39;</span><span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/huARdb_TCR_DeepInsight_notebook_38_0.png" src="_images/huARdb_TCR_DeepInsight_notebook_38_0.png" />
</div>
</div>
</div>
<div class="section" id="cluster-tcrs-in-reference-datasets">
<h2>Cluster TCRs in reference datasets<a class="headerlink" href="#cluster-tcrs-in-reference-datasets" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tdi</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">update_anndata</span><span class="p">(</span><span class="n">gex_reference_adata</span><span class="p">)</span>
<span class="n">tcr_reference_adata</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">unique_tcr_by_individual</span><span class="p">(</span><span class="n">gex_reference_adata</span><span class="p">,</span> <span class="n">label_key</span><span class="o">=</span><span class="s1">&#39;cell_type_1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fri May  5 21:11:40 2023	TCRDeepInsight: initializing dataset
Fri May  5 21:11:40 2023	TCRDeepInsight: adding &#39;tcr&#39; to adata.obs
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tdi</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">get_pretrained_tcr_embedding</span><span class="p">(</span>
    <span class="n">tcr_adata</span><span class="o">=</span><span class="n">tcr_reference_adata</span><span class="p">,</span>
    <span class="n">bert_config</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">get_human_config</span><span class="p">(),</span>
    <span class="n">checkpoint_path</span><span class="o">=</span><span class="s1">&#39;./tcr_deep_insight/data/pretrained_weights/human_bert_tcr_768_v1.ckpt&#39;</span><span class="p">,</span>
    <span class="n">pca_path</span><span class="o">=</span><span class="s1">&#39;./tcr_deep_insight/data/pretrained_weights/human_bert_tcr_pca_v1.pkl&#39;</span><span class="p">,</span>
    <span class="n">use_pca</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fri May  5 21:12:45 2023	Building BERT model
Fri May  5 21:12:47 2023	Loading BERT model checkpoints...
Fri May  5 21:12:47 2023	Computing TCR Embeddings...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|██████████████████████████████████████████████████████████| 156/156 [12:01&lt;00:00,  4.62s/it]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s1">&#39;./tcr_deep_insight/data/reference/human_tcr_reference_v1.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="alert alert-info">
<h2> Note </h2>
<p>You should have faiss-gpu installed to set <code class="docutils literal notranslate"><span class="pre">gpu=1</span></code>. This is much faster than the CPU version. For more information, please see <a class="reference external" href="https://pypi.org/project/faiss-gpu/">https://pypi.org/project/faiss-gpu/</a>.</p>
</div><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_cluster_result</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">cluster_tcr</span><span class="p">(</span>
    <span class="n">tcr_reference_adata</span><span class="p">,</span>
    <span class="n">label_key</span><span class="o">=</span><span class="s1">&#39;disease_type_1&#39;</span><span class="p">,</span>
    <span class="n">gpu</span><span class="o">=</span><span class="mi">1</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING clustering 635393 points to 635393 centroids: please provide at least 24780327 training points
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of training points (635393) same as number of clusters, just copying
Fri May  5 20:33:37 2023	Iterative anchoring...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████████████████████████████████████████| 635393/635393 [03:03&lt;00:00, 3464.95it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fri May  5 20:36:47 2023	Remove repetitive clusters...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████████████████████████████████████████| 170350/170350 [00:22&lt;00:00, 7429.55it/s]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>


<span class="n">result_tcr</span> <span class="o">=</span> <span class="n">tcr_cluster_result</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span>
    <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tcr_cluster_result</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;disease_type_1&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="s1">&#39;Solid tumor&#39;</span><span class="p">)</span> <span class="o">&amp;</span> 
    <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tcr_cluster_result</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">2</span><span class="p">)</span> <span class="o">&amp;</span>
    <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tcr_cluster_result</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;number_of_individuals&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">)</span>
<span class="p">]</span>


<span class="n">fig</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">createFig</span><span class="p">((</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">ax</span><span class="o">.</span><span class="n">scatter</span><span class="p">(</span>
    <span class="n">result_tcr</span><span class="p">[</span><span class="s1">&#39;tcr_similarity_score&#39;</span><span class="p">],</span>
    <span class="n">result_tcr</span><span class="p">[</span><span class="s1">&#39;disease_specificity_score&#39;</span><span class="p">],</span>
    <span class="n">s</span><span class="o">=</span><span class="n">result_tcr</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">*</span> <span class="mi">6</span><span class="p">,</span> 
    <span class="n">linewidths</span><span class="o">=</span><span class="mi">0</span>
<span class="p">)</span>
<span class="n">plt</span><span class="o">.</span><span class="n">show</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/huARdb_TCR_DeepInsight_notebook_45_0.png" src="_images/huARdb_TCR_DeepInsight_notebook_45_0.png" />
</div>
</div>
<div class="alert alert-info">
<h2> Note </h2>
<p>You should have <code class="docutils literal notranslate"><span class="pre">mafft</span></code> installed in your system to produce the logoplot below.</p>
<p>In <strong>Ubuntu</strong>,</p>
<p><code class="docutils literal notranslate"><span class="pre">sudo</span> <span class="pre">apt</span> <span class="pre">update</span></code>
<code class="docutils literal notranslate"><span class="pre">sudo</span> <span class="pre">apt</span> <span class="pre">install</span> <span class="pre">mafft</span></code></p>
<p>In <strong>MacOS</strong>, please use HomeBrew</p>
<p><code class="docutils literal notranslate"><span class="pre">brew</span> <span class="pre">install</span> <span class="pre">mafft</span></code></p>
<p>In CentOS or other operating systems, please check In <strong>CentOS</strong>, please check <a class="reference external" href="https://mafft.cbrc.jp/alignment/software/linuxportable.html">https://mafft.cbrc.jp/alignment/software/linuxportable.html</a> for manual installation.</p>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcrs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">result_tcr</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;disease_specificity_score&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">41</span><span class="p">]))</span> 

<span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_gex_selected_tcrs</span><span class="p">(</span>
    <span class="n">gex_reference_adata</span><span class="p">,</span>
    <span class="n">tcrs</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cell_type_1&#39;</span><span class="p">,</span>
    <span class="n">palette</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palettes</span><span class="o">.</span><span class="n">reannotated_prediction_palette</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Text(0.5, 1.0, &#39;CDR3b&#39;)
</pre></div>
</div>
<img alt="_images/huARdb_TCR_DeepInsight_notebook_47_1.png" src="_images/huARdb_TCR_DeepInsight_notebook_47_1.png" />
</div>
</div>
</div>
</div>
<div class="tex2jax_ignore mathjax_ignore section" id="cluster-tcrs-in-querying-datasets-from-reference-datasets">
<h1>Cluster TCRs in querying datasets from reference datasets<a class="headerlink" href="#cluster-tcrs-in-querying-datasets-from-reference-datasets" title="Permalink to this headline">¶</a></h1>
<p>Once the single-cell immune profiling datasets are processed by CellRanger and the GEX and TCR information are integrated by Scanpy, and Scirpy, you would get a datasets including:</p>
<ol class="simple">
<li><p>The raw gene expression matrix</p></li>
<li><p>The Full-length TCR sequence for each single-cell</p></li>
</ol>
<p>And you would provide the sample name as well as the individual name to the dataset</p>
<p>We use an example datasets from Sun et al., 2022 of Gastric Cancer (<a class="reference external" href="https://doi.org/10.1038/s41467-022-32627-z">Sun, K. et al. (2022)</a>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gex_query_adata</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;./tcr_deep_insight/data/GEX.GC.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gex_query_adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>IR_VJ_1_junction_aa</th>
      <th>IR_VDJ_1_junction_aa</th>
      <th>IR_VJ_1_v_call</th>
      <th>IR_VJ_1_j_call</th>
      <th>IR_VDJ_1_v_call</th>
      <th>IR_VDJ_1_j_call</th>
      <th>sample_name</th>
      <th>individual_name</th>
      <th>reannotated_prediction_3</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAACCTGTCACAGTAC-Sun_2022-GC05P</th>
      <td>CALSRYSGAGSYQLTF</td>
      <td>CASSLPLANTIYF</td>
      <td>TRAV16</td>
      <td>TRAJ28</td>
      <td>TRBV7-9</td>
      <td>TRBJ1-3</td>
      <td>GC05P</td>
      <td>GC05</td>
      <td>Treg</td>
    </tr>
    <tr>
      <th>AAACCTGTCTTGCAAG-Sun_2022-GC05P</th>
      <td>CAVISSDNARLMF</td>
      <td>CSVGRDQKNIQYF</td>
      <td>TRAV36DV7</td>
      <td>TRAJ31</td>
      <td>TRBV29-1</td>
      <td>TRBJ2-4</td>
      <td>GC05P</td>
      <td>GC05</td>
      <td>CD4</td>
    </tr>
    <tr>
      <th>AAACGGGTCAGATAAG-Sun_2022-GC05P</th>
      <td>CAVERDSGGSNYKLTF</td>
      <td>CASSQRQGFNEQFF</td>
      <td>TRAV22</td>
      <td>TRAJ53</td>
      <td>TRBV4-1</td>
      <td>TRBJ2-1</td>
      <td>GC05P</td>
      <td>GC05</td>
      <td>CD8</td>
    </tr>
    <tr>
      <th>AAAGATGAGGTTACCT-Sun_2022-GC05P</th>
      <td>CAATPGADKLIF</td>
      <td>CASSRTGSLYGYTF</td>
      <td>TRAV29DV5</td>
      <td>TRAJ34</td>
      <td>TRBV27</td>
      <td>TRBJ1-2</td>
      <td>GC05P</td>
      <td>GC05</td>
      <td>Treg</td>
    </tr>
    <tr>
      <th>AAAGATGCAGGGTACA-Sun_2022-GC05P</th>
      <td>CAGETSGSARQLTF</td>
      <td>CASSFQSYGRDKF</td>
      <td>TRAV12-3</td>
      <td>TRAJ22</td>
      <td>TRBV5-4</td>
      <td>TRBJ1-2</td>
      <td>GC05P</td>
      <td>GC05</td>
      <td>CD4</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TTTGCGCGTTGGACCC-Sun_2022-GC04P</th>
      <td>CAMTSVSSGTYKYIF</td>
      <td>CASSLQDFNTGELFF</td>
      <td>TRAV14DV4</td>
      <td>TRAJ40</td>
      <td>TRBV7-9</td>
      <td>TRBJ2-2</td>
      <td>GC04P</td>
      <td>GC04</td>
      <td>CD8</td>
    </tr>
    <tr>
      <th>TTTGCGCTCGGCGCTA-Sun_2022-GC04P</th>
      <td>CVVDTGRRALTF</td>
      <td>CASSPGQGAYEQYF</td>
      <td>TRAV12-1</td>
      <td>TRAJ5</td>
      <td>TRBV7-3</td>
      <td>TRBJ2-7</td>
      <td>GC04P</td>
      <td>GC04</td>
      <td>MAIT</td>
    </tr>
    <tr>
      <th>TTTGGTTCACTGCCAG-Sun_2022-GC04P</th>
      <td>CAVRTSKLTF</td>
      <td>CSAWTSGSSYNEQFF</td>
      <td>TRAV21</td>
      <td>TRAJ44</td>
      <td>TRBV20-1</td>
      <td>TRBJ2-1</td>
      <td>GC04P</td>
      <td>GC04</td>
      <td>CD4</td>
    </tr>
    <tr>
      <th>TTTGGTTGTAAATACG-Sun_2022-GC04P</th>
      <td>CAMTSVSSGTYKYIF</td>
      <td>CASSLQDFNTGELFF</td>
      <td>TRAV14DV4</td>
      <td>TRAJ40</td>
      <td>TRBV7-9</td>
      <td>TRBJ2-2</td>
      <td>GC04P</td>
      <td>GC04</td>
      <td>CD8</td>
    </tr>
    <tr>
      <th>TTTGTCACAGAGCCAA-Sun_2022-GC04P</th>
      <td>CIVGNTDKLIF</td>
      <td>CASSQGSGNTIYF</td>
      <td>TRAV26-1</td>
      <td>TRAJ34</td>
      <td>TRBV11-3</td>
      <td>TRBJ1-3</td>
      <td>GC04P</td>
      <td>GC04</td>
      <td>CD4</td>
    </tr>
  </tbody>
</table>
<p>59417 rows × 9 columns</p>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gex_query_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;individual&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">map</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;-&quot;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="n">gex_query_adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">index</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tdi</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">update_anndata</span><span class="p">(</span><span class="n">gex_query_adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fri May  5 20:09:15 2023	TCRDeepInsight: initializing dataset
Fri May  5 20:09:15 2023	TCRDeepInsight: adding &#39;tcr&#39; to adata.obs
</pre></div>
</div>
</div>
</div>
<p>We wrapped our TCR-DeepInsight model into few lines of Python code here</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">_</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">get_pretrained_gex_embedding</span><span class="p">(</span>
    <span class="n">gex_query_adata</span><span class="p">,</span>
    <span class="n">gex_reference_adata</span><span class="o">=</span><span class="n">gex_reference_adata</span><span class="p">,</span>
    <span class="n">transfer_label</span><span class="o">=</span><span class="s1">&#39;cell_type_1&#39;</span><span class="p">,</span>
    <span class="n">checkpoint_path</span><span class="o">=</span><span class="s1">&#39;./tcr_deep_insight/data/pretrained_weights/human_vae_gex_all_cd4_cd8_v1.ckpt&#39;</span><span class="p">,</span>
    <span class="n">device</span><span class="o">=</span><span class="s1">&#39;cuda:0&#39;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fri May  5 20:09:30 2023	Loading Reference adata...
Fri May  5 20:09:39 2023	Building VAE model...
Fri May  5 20:09:39 2023	Initializing dataset into memory
Fri May  5 20:09:39 2023	warning: the provided n_batch=0 does not match the number of batch in the adata.
Fri May  5 20:09:39 2023	setting n_batch to 437
Fri May  5 20:10:00 2023	Loading VAE model checkpoints...
Fri May  5 20:10:02 2023	Initializing dataset into memory
Fri May  5 20:10:02 2023	warning: the provided n_batch=437 does not match the number of batch in the adata.
Fri May  5 20:10:02 2023	setting n_batch to 461
Fri May  5 20:10:57 2023	Transfering cell type labels...
</pre></div>
</div>
</div>
</div>
<p>We get the joint representation of the reference GEX and the query new dataset GEX and project then to 2D using UMAP</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="n">fig</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">createFig</span><span class="p">((</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">gex_query_adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cell_type_1&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palette</span><span class="o">.</span><span class="n">reannotated_prediction_palette</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>findfont: Font family [&#39;Arial&#39;] not found. Falling back to DejaVu Sans.
findfont: Font family [&#39;Arial&#39;] not found. Falling back to DejaVu Sans.
findfont: Font family [&#39;Arial&#39;] not found. Falling back to DejaVu Sans.
</pre></div>
</div>
<img alt="_images/huARdb_TCR_DeepInsight_notebook_58_1.png" src="_images/huARdb_TCR_DeepInsight_notebook_58_1.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="n">ig</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">createFig</span><span class="p">((</span><span class="mi">5</span><span class="p">,</span><span class="mi">5</span><span class="p">))</span>
<span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">gex_query_adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s1">&#39;sample_name&#39;</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palettes</span><span class="o">.</span><span class="n">godsnot_102</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/huARdb_TCR_DeepInsight_notebook_59_0.png" src="_images/huARdb_TCR_DeepInsight_notebook_59_0.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_query_adata</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">unique_tcr_by_individual</span><span class="p">(</span><span class="n">gex_query_adata</span><span class="p">,</span> <span class="n">label_key</span><span class="o">=</span><span class="s1">&#39;cell_type_1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_query_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;disease_type_2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Gastric Cancer&#39;</span>
<span class="n">tcr_reference_adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;disease_type_2&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;Others&#39;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tdi</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">get_pretrained_tcr_embedding</span><span class="p">(</span>
    <span class="n">tcr_adata</span><span class="o">=</span><span class="n">tcr_query_adata</span><span class="p">,</span>
    <span class="n">bert_config</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">config</span><span class="o">.</span><span class="n">get_human_config</span><span class="p">(),</span>
    <span class="n">checkpoint_path</span><span class="o">=</span><span class="s1">&#39;./tcr_deep_insight/data/pretrained_weights/human_bert_tcr_768_v1.ckpt&#39;</span><span class="p">,</span>
    <span class="n">pca_path</span><span class="o">=</span><span class="s1">&#39;./tcr_deep_insight/data/pretrained_weights/human_bert_tcr_pca_v1.pkl&#39;</span><span class="p">,</span>
    <span class="n">use_pca</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fri May  5 20:40:52 2023	Building BERT model
Fri May  5 20:40:53 2023	Loading BERT model checkpoints...
Fri May  5 20:40:53 2023	Computing TCR Embeddings...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████████████████████████████████████████████████████████████| 9/9 [00:40&lt;00:00,  4.55s/it]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcr_cluster_result</span> <span class="o">=</span> <span class="n">tdi</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">cluster_tcr_from_reference</span><span class="p">(</span>
    <span class="n">tcr_query_adata</span><span class="p">,</span>
    <span class="n">tcr_reference_adata</span><span class="p">,</span>
    <span class="n">label_key</span><span class="o">=</span><span class="s1">&#39;disease_type_2&#39;</span><span class="p">,</span>
    <span class="n">gpu</span><span class="o">=</span><span class="mi">1</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>WARNING clustering 671740 points to 671740 centroids: please provide at least 26197860 training points
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Number of training points (671740) same as number of clusters, just copying
Fri May  5 21:27:19 2023	Iterative anchoring...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|███████████████████████████████████████████████████████████████████████████████████| 36347/36347 [00:10&lt;00:00, 3563.76it/s]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Fri May  5 21:27:29 2023	Remove repetitive clusters...
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>100%|████████████████████████████████████████████████████████████████████████████████████| 8471/8471 [00:01&lt;00:00, 7507.02it/s]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tdi</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">inject_tcr_cluster</span><span class="p">(</span><span class="n">tcr_query_adata</span><span class="p">,</span> <span class="n">tcr_cluster_result</span><span class="p">,</span> <span class="s1">&#39;cell_type_1&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="n">result_tcr</span> <span class="o">=</span> <span class="n">tcr_cluster_result</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span>
    <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tcr_cluster_result</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;count&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">)</span> <span class="o">&amp;</span> 
    <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tcr_cluster_result</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;number_of_individuals&#39;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">)</span> <span class="o">&amp;</span> 
    <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">tcr_cluster_result</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;cell_type_1&#39;</span><span class="p">]</span> <span class="o">!=</span> <span class="s1">&#39;MAIT&#39;</span><span class="p">)</span>
<span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">gex_query_adata</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s2">&quot;/Users/snow/Downloads/gex_query_adata.h5ad&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">tcrs</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">filter</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">x</span> <span class="o">!=</span> <span class="s1">&#39;-&#39;</span><span class="p">,</span> <span class="n">result_tcr</span><span class="o">.</span><span class="n">sort_values</span><span class="p">(</span><span class="s1">&#39;tcr_similarity_score&#39;</span><span class="p">,</span> <span class="n">ascending</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span><span class="mi">1</span><span class="p">:</span><span class="mi">41</span><span class="p">]))</span> 

<span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">plot_gex_tcr_selected_tcrs</span><span class="p">(</span>
    <span class="n">gex_adata</span> <span class="o">=</span> <span class="n">gex_query_adata</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span>
    <span class="n">tcrs</span><span class="o">=</span><span class="n">tcrs</span><span class="p">,</span>
    <span class="n">palette</span><span class="o">=</span><span class="n">tdi</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">palette</span><span class="o">.</span><span class="n">reannotated_prediction_palette</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="_images/huARdb_TCR_DeepInsight_notebook_67_0.png" src="_images/huARdb_TCR_DeepInsight_notebook_67_0.png" />
</div>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./."
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
            
                <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="huARdb_TCR_paper_notebook.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Import Basic Packages and Definitions</p>
        </div>
    </a>
</div>
            
        </div>
    </div>
    <footer class="footer">
  <p>
    
      By Ziwei Xue<br/>
    
        &copy; Copyright 2023.<br/>
  </p>
</footer>
</main>


      </div>
    </div>
  
  <script src="_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>